### PR TITLE
[#7537] Add `--ignore-symlinks` alias for `--link` option (main)

### PR DESCRIPTION
**Pull request by korydraughn**
Not yet tested.

### CODE REVIEW

**Comment by trel** [comment]
make sure to create an issue for removing the --link eventually as well

File: lib/core/src/irods_parse_command_line_options.cpp [hunk 1]
```diff
@@ -26,38 +26,41 @@ static int parse_program_options(
...
+        ("rbudp,Q", "use RBUDP (datagram) protocol for the data transfer")
+        ("recursive,r", "recursive - store the whole subdirectory")
+        ("dest_resc,R", po::value<std::string>(), "resource - specifies the resource to store to. This can also be specified in your environment or via a rule set up by the administrator")
+        ("ticket,t", po::value<std::string>(), "ticket - ticket (string) to use for ticket-based access")
+        ("renew_socket,T", "renew socket connection after 10 minutes")
+        ("verbose,v", "verbose")
+        ("very_verbose,V", "very verbose")
+        ("data_type,D", po::value<std::string>(), "dataType - the data type string")
+        ("restart_file,X", po::value<std::string>(), "restartFile - specifies that the restart option is on and the restartFile input specifies a local file that contains the restart information.")
+        ("link", "ignore symlink.")
```
Comment (trel) [inline lib/core/src/irods_parse_command_line_options.cpp:52]: could make this plural

Comment (korydraughn) [reply lib/core/src/irods_parse_command_line_options.cpp:52]: Missed that one.

Comment (trel) [reply lib/core/src/irods_parse_command_line_options.cpp:52]: just the description

File: scripts/irods/test/test_iput.py [hunk 2]
```diff
@@ -821,6 +821,25 @@ def test_iput_does_not_ignore_symlinks_by_default__issue_5359(self):
...
+        with open(test_file, 'w') as f:
+            f.write('This file will be ignored by iput due to the symlink.')
+
+        # Create a symlink to the file.
+        symlinked_file = f'{self.user.local_session_dir}/{test_file_basename}.symlinked'
+        os.symlink(test_file, symlinked_file)
+        self.assertTrue(os.path.exists(symlinked_file))
+
+        # Try to upload the symlinked file into iRODS.
+        self.user.assert_icommand(['iput', '--ignore-symlinks', symlinked_file])
```
Comment (trel) [inline scripts/irods/test/test_iput.py:837]: please try a recursive put as well

File: lib/core/src/irods_parse_command_line_options.cpp (hunk 1 above)
Comment (korydraughn) [reply lib/core/src/irods_parse_command_line_options.cpp]: Yes, the icommands were already updated to use the plural form for the deprecated option. I just missed this one.

Comment (korydraughn) [reply lib/core/src/irods_parse_command_line_options.cpp]: Done.

**Review by alanking — COMMENTED** [review]
`irsync` uses `parseCmdLineOpt`, not `parse_program_options`, so that one will need to be updated as well.

File: lib/core/src/irods_parse_command_line_options.cpp [hunk 3]
```diff
@@ -26,38 +26,41 @@ static int parse_program_options(
...
+        ("rbudp,Q", "use RBUDP (datagram) protocol for the data transfer")
+        ("recursive,r", "recursive - store the whole subdirectory")
+        ("dest_resc,R", po::value<std::string>(), "resource - specifies the resource to store to. This can also be specified in your environment or via a rule set up by the administrator")
+        ("ticket,t", po::value<std::string>(), "ticket - ticket (string) to use for ticket-based access")
+        ("renew_socket,T", "renew socket connection after 10 minutes")
+        ("verbose,v", "verbose")
+        ("very_verbose,V", "very verbose")
+        ("data_type,D", po::value<std::string>(), "dataType - the data type string")
+        ("restart_file,X", po::value<std::string>(), "restartFile - specifies that the restart option is on and the restartFile input specifies a local file that contains the restart information.")
+        ("link", "ignore symlinks.")
```
[the remark this answers is not in the document]
Comment (alanking) [reply lib/core/src/irods_parse_command_line_options.cpp:52]: On second thought, does this description actually show up anywhere? Well, anyway, I think we should mark it here as well.

File: scripts/irods/test/test_iput.py [hunk 4]
```diff
@@ -821,6 +821,25 @@ def test_iput_does_not_ignore_symlinks_by_default__issue_5359(self):
...
 
+    def test_iput_ignore_symlinks_option_is_an_alias_of_the_link_option__issue_7537(self):
+        # Create a file.
+        test_file_basename = 'iput_issue_7537.txt'
+        test_file = f'{self.user.local_session_dir}/{test_file_basename}'
+        with open(test_file, 'w') as f:
+            f.write('This file will be ignored by iput due to the symlink.')
+
+        # Create a symlink to the file.
+        symlinked_file = f'{self.user.local_session_dir}/{test_file_basename}.symlinked'
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply scripts/irods/test/test_iput.py:832]: Oh I see how that can be confusing.

Will fix.

File: lib/core/src/irods_parse_command_line_options.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (korydraughn) [reply lib/core/src/irods_parse_command_line_options.cpp:52]: Will do.

[the remark this answers is not in the document]
Comment (korydraughn) [reply lib/core/src/irods_parse_command_line_options.cpp]: Done.

File: scripts/irods/test/test_iput.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (korydraughn) [reply scripts/irods/test/test_iput.py]: Done.

File: scripts/irods/test/test_iput.py (hunk 2 above)
Comment (korydraughn) [reply scripts/irods/test/test_iput.py]: Done.

**Comment by korydraughn** [comment]
All unit tests passed.

Running core tests.

File: lib/core/src/irods_parse_command_line_options.cpp [hunk 5]
```diff
@@ -26,38 +26,41 @@ static int parse_program_options(
...
+        ("rbudp,Q", "use RBUDP (datagram) protocol for the data transfer")
+        ("recursive,r", "recursive - store the whole subdirectory")
+        ("dest_resc,R", po::value<std::string>(), "resource - specifies the resource to store to. This can also be specified in your environment or via a rule set up by the administrator")
+        ("ticket,t", po::value<std::string>(), "ticket - ticket (string) to use for ticket-based access")
+        ("renew_socket,T", "renew socket connection after 10 minutes")
+        ("verbose,v", "verbose")
+        ("very_verbose,V", "very verbose")
+        ("data_type,D", po::value<std::string>(), "dataType - the data type string")
+        ("restart_file,X", po::value<std::string>(), "restartFile - specifies that the restart option is on and the restartFile input specifies a local file that contains the restart information.")
+        ("link", "[Deprecated] ignore symlinks.")
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply lib/core/src/irods_parse_command_line_options.cpp:52]: Will do.

File: scripts/irods/test/test_irsync.py [hunk 6]
```diff
@@ -646,3 +646,22 @@ def test_irsync_does_not_ignore_symlinks_by_default__issue_5359(self):
...
+        symlink_file = f'{self.user0.local_session_dir}/{test_file_basename}.symlink'
+        os.symlink(test_file, symlink_file)
+        self.assertTrue(os.path.exists(symlink_file))
+
+        # Try to upload the symlink file into iRODS.
+        logical_path = f'{self.user0.session_collection}/{test_file_basename}'
+        self.user0.assert_icommand(['irsync', '--ignore-symlinks', symlink_file, f'i:{logical_path}'])
+
+        # Show no data object exists, proving the symlink file was ignored.
+        self.assertFalse(lib.replica_exists(self.user0, logical_path, 0))
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply scripts/irods/test/test_irsync.py:667]: Will do.

**Comment by korydraughn** [comment]
All core tests passed.

**Comment by korydraughn** [comment]
Will rerun the irsync test after https://github.com/irods/irods/pull/7567#discussion_r1527185234 is resolved.

File: lib/core/src/irods_parse_command_line_options.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (korydraughn) [reply lib/core/src/irods_parse_command_line_options.cpp]: Done.

**Comment by korydraughn** [comment]
Fixed a couple issues with the tests.

This PR is ready now.

**Review by alanking — COMMENTED** [review]
Makes sense to me. Squash to taste

**Comment by korydraughn** [comment]
@alanking So it's correct that if symlinks are followed, the data object that appears in the catalog will share the same name as the symlink, not the symlinked file?

**Comment by alanking** [comment]
That's correct. Here's the behavior in 4.3.1 as released...

The setup:
```bash
$ ls -l linkz/
total 8
-rw-rw-r-- 1 irods irods    5 Mar 18 16:54 somefile
drwxrwxr-x 2 irods irods 4096 Mar 18 16:54 subdir
$ ls -l linkz/subdir/ 
total 0
lrwxrwxrwx 1 irods irods 11 Mar 18 16:54 this_is_a_link -> ../somefile
$ cat linkz/somefile  
blah
$ cat linkz/subdir/this_is_a_link 
blah
```

`iput` with no `--link` option:
```bash
$ iput -r linkz linkz_no_ignore_symlinks
Running recursive pre-scan... pre-scan complete... transferring data...
$ # there's the regular file
$ ils -l linkz_no_ignore_symlinks
/tempZone/home/rods/linkz_no_ignore_symlinks:
  rods              0 demoResc            5 2024-03-18.16:55 & somefile
  C- /tempZone/home/rods/linkz_no_ignore_symlinks/subdir  
$ # and there's the symlink with its symlink name (not the target file)
$ ils -l linkz_no_ignore_symlinks/subdir
/tempZone/home/rods/linkz_no_ignore_symlinks/subdir:
  rods              0 demoResc            5 2024-03-18.16:55 & this_is_a_link
```
`iput` with `--link` option:
```bash
$ iput --link -r linkz linkz_ignore_symlinks
Running recursive pre-scan... pre-scan complete... transferring data...
$ # there's the regular file
$ ils -l linkz_ignore_symlinks       
/tempZone/home/rods/linkz_ignore_symlinks:
  rods              0 demoResc            5 2024-03-18.16:56 & somefile
  C- /tempZone/home/rods/linkz_ignore_symlinks/subdir  
$ # there's no symlink in here because it was ignored
$ ils -l linkz_ignore_symlinks/subdir
/tempZone/home/rods/linkz_ignore_symlinks/subdir:
```

**Comment by trel** [comment]
this is correct and expected.

**Comment by korydraughn** [comment]
Good. Just wanted to confirm everyone's thoughts aligned with what I saw while updating the tests.

Will pound soon.

**Comment by korydraughn** [comment]
#'d

**Review by alanking — APPROVED** [review]

**Comment by alanking** [comment]
Will merge this and https://github.com/irods/irods_client_icommands/pull/449 together.

**Pull request merged**